lewis/adapters/modbus.py

```diff
@@ -0,0 +1,569 @@
+# -*- coding: utf-8 -*-
+# *********************************************************************
+# lewis - a library for creating hardware device simulators
+# Copyright (C) 2016 European Spallation Source ERIC
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+# References Used:
+#   http://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
+#   http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf
+#   https://github.com/sourceperl/pyModbusTCP
+#   https://github.com/bashwork/pymodbus
+# *********************************************************************
+
+
+from __future__ import division
+
+import socket
+import asyncore
+import struct
+
+from copy import deepcopy
+from math import ceil
+from argparse import ArgumentParser
+
+from lewis.adapters import Adapter
+
+
+class ModbusDataBank(object):
+    def __init__(self, config):
+        """
+        Preliminary DataBank implementation for Modbus.
+
+        This is a very generic implementation of a databank for Modbus. It's meant to set the
+        groundwork for future implementations. The create_basic method should be used for
+        construction, rather than this initializer. The signature of this __init__ method is
+        subject to change.
+
+        .. sourcecode:: Python
+
+            di = ModbusDataBank.create_basic(False, 0x1000, 0x1FFF)
+
+        :param config: dict containing configuration
+        """
+        self._data = config['data']
+        self._start_addr = config['start_addr']
+
+    @classmethod
+    def create_basic(cls, default_value=0, start_addr=0x0000, last_addr=0xFFFF):
```

Comment:
> The internal implementation of ModbusDataBank (including get and set) would of course have to be extended significantly to support all this.

This sounds like an argument for sub-classing to me. I like your idea of using `kwargs` for the base `__init__`, it might simplify change in the future as well.

So how about something like this?

```python
class ModbusDataBank(object):
    def __init__(self, *args, **kwargs):
        self.data = kwargs['data']
        self._start_addr = kwargs['start_addr']

    # and so on

class ModbusBasicDataBank(ModbusDataBank):
    def __init__(default_value=0, start_addr=0x0000, last_addr=0xFFFF):
        super(ModbusBasicDataBank, self).__init__(start_addr=start_addr, data=[default_value] * (last_addr - start_addr + 1))
```

That way `ModbusDataBank` does not need to change when new implementations are added. The downside is you'd need to know what sub-classes exist in order to import them to implement your interface, but this way it's also easier to draw a line and say "don't use this directly, use one of the sub-classes instead".